added a management command to output versioning rule violation resour…

hyi opened the pull request:
…ce list

<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [ ] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. [Enter positive test case here]

hyi commented:
@dtarb I wrote a new management command to fix versioning rule violation resources based on the list you worked out, and ran the new management command on beta, then ran the management command on beta to output versioning rule violation resources afterwards, and can confirm all issues with isVersionOf and isReplacedBy not in pairs have been fixed. We still have 6 resources being replaced by multiple resources. The output file that includes these remaining versioning violations can be downloaded here:
[is_replaced_by_violations2.csv](https://github.com/hydroshare/hydroshare/files/7530182/is_replaced_by_violations2.csv)
 I found these resources are not in the list your worked out, so it'd be great if you could take a look at these 6 resources and see what is going on. Hopefully this will not take much of your time. Thanks

dtarb commented:
@hyi all the resources in is_replaced_by_violations2.csv belong to Tony Castronova with the "Test - REMOVE ME"  I labeled them in the OtherCases tab as Tony test - can delete.  I prefer to just delete them rather than try to fix them.  If @Castronova disagrees - then he can take the time to sort this out.

I identified the problem with the Christina resources.  See new tab Christina in the spreadsheet, and the suggested replacements to fix them.

dtarb commented:
An alternative way to resolve the Tony test cases if you can not delete them (which may not be possible given the versioning delete rules) is just to delete all their versioning relationships.  Then Tony will be left with 6 resources with the same title that he can remove at his leisure.

hyi commented:
@dtarb I fixed those problematic resources owned by Christina and Tony as you suggested and ran the fix on beta, then ran the management command to output problematic resource lists on beta. Although the management command did not detect any problematic resources after running the fix, I noticed the number of version of relations is not equal to the number of replaced by relations, which is not right. After looking into the resources and my code, I realized I missed to check a couple of additional condition in order to detect all the cases when the two relations are not in pairs. I fixed the management command and it now outputs additional resource relations that violate the versioning chain rule. The output is uploaded here
[replace_by_and_version_of_not_in_pairs.csv](https://github.com/hydroshare/hydroshare/files/7541610/replace_by_and_version_of_not_in_pairs.csv)
: 
I believe this will be the last fix we need to make to fix all problematic resources. I will update the input list to handle these few additional resources and run it on beta after I receive your updated list to handle those captured in this new csv file.

hyi commented:
@dtarb I incorporated your new rows in all change list and ran the versioning fix on beta. Now the not-in-pair problematic resources are reduced from the initial 25 to 21. The 21 items in the latest not-in-pair csv file is pasted below for easy reference:
```
  1 Resource,isReplacedBy,isVersionOf
  2 7cf9104f166e43b7aa932e9e9dc855e3,0efc1103b5ff4ca3bee04b70450d9e98,
  3 0773143f6ddd43b780d2837109280ade,005d6cfd91be4c448cc64394c10f8397,
  4 2853f728f5024f27b4ee8ca929f8bf58,,2d9787bf36d04c9383e595d179f9298b
  5 b9a8878e65cc4345912ca203b0b5143c,,ea0952c7a16b42468822a78db3de5a4b
  6 4111306529e74503b090494ef1e808e2,,fdc3a06e6ad842abacfa5b896df73a76
  7 224556101f3641109910595b4a761276,,dea233478adf45ee8f71fb878828951e
  8 c487ef1251ef41668c9cd36cd5630f05,,dea233478adf45ee8f71fb878828951e
  9 973dec7c8d504ab89968a4fa01f12764,,d7b7f9d5c7e0411683b76ef45c6bf0d8
 10 6c0217a019ba47a585221d305fc09030,,d7b7f9d5c7e0411683b76ef45c6bf0d8
 11 1f24c2a5268a4c919760f1238cc7f2ed,,d7b7f9d5c7e0411683b76ef45c6bf0d8
 12 3264343d39a84292b8411d213748aa99,,d7b7f9d5c7e0411683b76ef45c6bf0d8
 13 21970a8943d6426ea8c39df1dc7c1861,,d7b7f9d5c7e0411683b76ef45c6bf0d8
 14 aa77895efa764ad0a4b2a40d18af23f6,,d7b7f9d5c7e0411683b76ef45c6bf0d8
 15 0efc1103b5ff4ca3bee04b70450d9e98,,7f9ba6e8bfca4bf1b9395667de0ff0e5
 16 28e3bb9c42ec4fca9b841afbfa660764,,dea233478adf45ee8f71fb878828951e
 17 005d6cfd91be4c448cc64394c10f8397,,b15ecfb0a5e946e38c0f49646fbc33bb
 18 4763446ede6344e4ab5c5b1790631a69,,abb61157105746e5a03f983c0c6a8249
 19 6f3974296719464ea7882570e549975f,,73c4f3dcff884a6da2c0982df769987c
 20 f06ae5edd79a4844a968ea224043b20d,,73c4f3dcff884a6da2c0982df769987c
 21 2f8ea9fe424e4d55992e8a71d9945e35,,873b5490f986428bbd23e37c81906402
 22 3ff80acf24894fb6b24ae4111e9a95ea,,873b5490f986428bbd23e37c81906402
```
I don't see a pattern so that we can fix these remaining resources by logic, so we will have to walk through these remaining resources manually to fix them.

dtarb commented:
@hyi I have updated https://docs.google.com/spreadsheets/d/1mSTOJbwqA_aPH8rmhpXrPkBK45pyAe5R/edit#gid=755760022 sheet "AllChangesList" to indicate fixes for the remaining problems.

Note, that in doing this, I posted two new issues #4449 which is cosmetic, and #4450 which addresses one of the root causes of these problems.  

In fixing these problems, note column D that indicates isDerivedFrom.  Some of the problems are fixed with derived from relationships.  

Some problems I fixed by deleting multiply versioned resources.  In doing this I noted a resource that appears incomplete or corrupt.  Noting this here to look in to it https://www.hydroshare.org/resource/517763712df44fd4b0903bf0bc9ab3b4/ and fix if you can, or file an issue once you understand it.

hyi commented:
@dtarb I applied your updated change list and tested on beta. Although more resources were fixed, there were still some problematic resources that don't have replaced by and version of relations in pair. I spent a few hours looking into these problematic resources in conjunction with others on your resource change list, and finally fixed all of them. I do have to add the following more rows to the curated list for fixing versioning issues: 
```
,222e832d3df24dea9bae9bbeb6f4219d,
,2d9787bf36d04c9383e595d179f9298b,
,cd0bc6f5a2e54247927128a85bc10c8a,
,973dec7c8d504ab89968a4fa01f12764,
,6c0217a019ba47a585221d305fc09030,
,1f24c2a5268a4c919760f1238cc7f2ed,
,3264343d39a84292b8411d213748aa99,
,21970a8943d6426ea8c39df1dc7c1861,
,3ff80acf24894fb6b24ae4111e9a95ea,
ea0952c7a16b42468822a78db3de5a4b,b832aebb732b447da7af0bf17ee21286,
```
With these more rows added, all resources with versioning issues are fixed on beta. I don't have edit permissions on the change list on the google sheet, so it'd be great if you can add these more rows into the end of the change list so that the list is up to date and reflect the latest implementation.

hyi commented:
@pkdash @sblack-usu it'd be great if one of you can do a quick review of this PR so that it can be merged for the next release. I have tested it on beta and all resources with versioning issues are fixed after running the management commands in this PR. Just need to have this merged so that they can be run on production to fix quite a few resources with versioning issues.

pkdash commented on hs_core/management/commands/fix_resource_versioning.py:
```diff
@@ -0,0 +1,117 @@
+                    # both isVersionOf and isReplacedby are empty, so clean up any versioning
+                    # relationships for this resource if any
+                    if res.metadata.relations.all().filter(type='isVersionOf').exists():
+                        res.metadata.relations.all().filter(type='isVersionOf').all().delete()
```
simpler version: `res.metadata.relations.filter(type='isVersionOf').delete()`

hyi replied:
@pkdash What if ```res.metadata.relations.filter(type='isVersionOf')``` is None? That would raise an exception with the simpler version statement, right? Checking whether the relation exists first will avoid this issue, right?

pkdash replied:
@hyi Even if a record doesn't exist it won't raise exception.

pkdash reviewed (changes requested):
I am wondering the need for 2 management commands to fix the resource versioning chain issue.   My thinking is as follows which can be done in one management command:
1. Delete all the orphaned (for which there is no resource exists) relation meta objects of type 'isVersionOf' and 'isReplacedBy'
2. For each resource keep the latest relation meta of type 'isVersionOf' and 'isReplacedBy' and delete all other duplicates of relation meta of these types. 

I am not aware of all the issues related to versioning chain, so my logic as described above may not work.

hyi commented on hs_core/management/commands/fix_resource_versioning.py:
```diff
@@ -0,0 +1,117 @@
+                        res.metadata.relations.all().filter(type='isReplacedBy').all().delete()
+                        resource_modified(res, overwrite_bag=False)
+                        fixed = True
+                    if res_id == '4111306529e74503b090494ef1e808e2':
```
This is one single case @dtarb pointed out in his change list that this missing copy relationship needs to be added. So I am just following Dave's suggestion to fix this resource automatically.

hyi commented on hs_core/management/commands/output_wrong_versioning_resources.py:
```diff
@@ -0,0 +1,105 @@
+           "that point to non-existant resources"
+
+    def handle(self, *args, **options):
+        replace_by_qs = Relation.objects.filter(type='isReplacedBy')
```
We are not deleting duplicate relation meta object here, but rather delete those relations if the resource id in the relations don't exist any more. Not sure how this happened, but we have such problems in the database. So the ordering does not matter in this type of cleanup.

hyi commented on hs_core/management/commands/output_wrong_versioning_resources.py:
```diff
@@ -0,0 +1,105 @@
+                replaced_by_dict[res_obj.short_id] = [replacy_by_id]
+
+        version_of_dict = {}
+        version_of_qs = Relation.objects.filter(type='isVersionOf')
```
See my response above - no need for ordering in this type of cleanup.

hyi commented on hs_core/management/commands/output_wrong_versioning_resources.py:
```diff
@@ -0,0 +1,105 @@
+                        writer.writerow([key, val])
+
+        # output resource list where isReplacedBy and isVersionOf are not in pairs
+        with open('replace_by_and_version_of_not_in_pairs.csv', 'w') as csvfile:
```
No, it is not used anywhere. It is just an output for admin's reference to see whether all resources have been fixed, i.e., not in pairs, after running fix_resource_versioning command.

pkdash replied:
Then where the csv file input for the command `fix_resource_versioning` comes from? Is that csv file created by Dave?

dtarb replied:
@pkdash I do not provide a CSV file.  @hyi and I worked out these details as discussed in #4028 and google sheet https://docs.google.com/spreadsheets/d/1mSTOJbwqA_aPH8rmhpXrPkBK45pyAe5R/edit#gid=755760022.  In this sheet the tab "AllChangesList" has the list of changes to enact, and I assume @hyi makes csv from this tab to be used by the command.

hyi commented:
@dtarb Yes, I have access to your change list now using my RENCI account. Thanks for your detailed comments. I'll look into them in detail and test on beta whenever I get a chance, probably tomorrow, since I am packed with meetings today.

hyi commented:
@dtarb I removed those items I added from the curated csv file, and tested it on beta again. I had to restore beta with the snapshot database first before testing since my previous tests already fixed all versioning chain issues. The output after I ran the management command to fix the versioning issues is shown below:
```
Resource,isReplacedBy,isVersionOf
973dec7c8d504ab89968a4fa01f12764,,d7b7f9d5c7e0411683b76ef45c6bf0d8
6c0217a019ba47a585221d305fc09030,,d7b7f9d5c7e0411683b76ef45c6bf0d8
1f24c2a5268a4c919760f1238cc7f2ed,,d7b7f9d5c7e0411683b76ef45c6bf0d8
3264343d39a84292b8411d213748aa99,,d7b7f9d5c7e0411683b76ef45c6bf0d8
21970a8943d6426ea8c39df1dc7c1861,,d7b7f9d5c7e0411683b76ef45c6bf0d8
3ff80acf24894fb6b24ae4111e9a95ea,,873b5490f986428bbd23e37c81906402
```
Since the first column of all resources in the output were supposed to be deleted on beta, these problematic output should be gone after these resources are deleted on beta to mimic the same state as the production. So I would claim the current curated list is a success to fix all resources with versioning issues.

hyi commented:
@dtarb I am uploading the curated csv file I created based on your all change list since it appears to fix all resources with versioning chain issues on beta as a reference.
[res_version_fix.csv](https://github.com/hydroshare/hydroshare/files/7857496/res_version_fix.csv)

hyi commented on hs_core/management/commands/fix_resource_versioning.py:
```diff
@@ -0,0 +1,117 @@
+                    except Exception:
+                        print(f'Resource {replace_res_id} does not exist', flush=True)
+                        continue
+                    if res.metadata.relations.all().filter(type='isVersionOf').exists():
```
I need this if statement so that resource modified can be applied and the counter for the number of fixed resources can be updated only when the relation exists.

hyi replied:
Never mind. I was thinking of a different statement. Yes, agreed this if statement here is not needed.

hyi commented:
@pkdash I have addressed or responded all of your code review comments. Take a look when you have a chance to see if everything looks good now.

pkdash reviewed (approved):
Left a comment for one more possible change.

hyi commented on hs_core/management/commands/fix_resource_versioning.py:
```diff
@@ -0,0 +1,117 @@
+                        hs_identifier = copy_of_res.metadata.identifiers.filter(name="hydroShareIdentifier").first()
+                        if hs_identifier:
+                            res.metadata.create_element('source', derived_from=hs_identifier.url)
+                        print(f'{res_id} is made a copy of {copy_of_res_id}', flush=True)
```
@pkdash Good catch. Have addressed this and made another commit. Will need you to approve again.

hyi commented:
@pkdash All checks have passed. Can you take a quick look again to approve the PR?

hyi commented:
@pkdash I have tested the management command for fixing versioning issues on beta after you deployed beta with your reference change work, and everything appears to work as expected. However, I did get these 2 line output from output remaining resources with versioning issues as shown below:
```
Resource,isReplacedBy,isVersionOf
cd0bc6f5a2e54247927128a85bc10c8a,,d7b7f9d5c7e0411683b76ef45c6bf0d8
9c4074af5f0b44c1ade1c0da7393c481,,8dbd221006f34c918f48c72a51d6d9e3
```
I checked these resources and looks like these resources were deleted on beta when I tested them last time. @dtarb Could you take a quick look of these couple of resources of yours and see if they can be deleted safely. If yes, we should be good without the need to update the curated input list to fix these resources.

dtarb commented:
@hyi @pkdash  The first row above is a legitimate versioning problem that I'd like to resolve by removing the isVersionOf relationship.  They are both my resources.  I will then delete the old one.  d7b7f9d5c7e0411683b76ef45c6bf0d8.  I've added a line (line 131) to the AllChangesList tab in google doc that does this.

The second one is the test case I developed to illustrate the problem reported in #4450.  I'll delete these resources once #4450 is closed, so you can ignore it being flagged by the script.

hyi commented:
@dtarb @pkdash I updated the input version fix list, reset beta with initial database, and run version fix command again with the updated input list. Now I only got one line output for replay by and version of relations not in pairs:
```
Resource,isReplacedBy,isVersionOf
9c4074af5f0b44c1ade1c0da7393c481,,8dbd221006f34c918f48c72a51d6d9e3
```
which is expected. So everything looks good and ready to run on production after the next release. I am uploading the latest version fix curated list here for reference.
[res_version_fix.csv](https://github.com/hydroshare/hydroshare/files/7926245/res_version_fix.csv)

hyi commented:
@dtarb Just a quick note that I had to add an addition row in the curated version resource fix input list in order to fix an additional row of not in pair relationship problem. The row added is copied below for reference:
```
,e648a56290bd4003bb0a158199dc61cb,37a7544cc3b944e88a8754bd6c1b8eb1
```
I am not sure why this problem did not happen on beta when I tested it out over a week ago, but adding this row fixed the versioning relationship not in pair issue and everything looks good except for your test resource. It is not ready for you to test yet since @pkdash's management command is yet to be run. Will let you know on slack when it is ready for you to test on production.
